examples/bsoncxx/builder_stream_fancy.cpp

```diff
@@ -0,0 +1,106 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <vector>
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/builder/stream/array.hpp>
+#include <bsoncxx/json.hpp>
+
+using namespace bsoncxx;
+
+// concatenates arbitrary ranges into an array context
+template <typename begin_t, typename end_t>
+class range_array_appender {
+   public:
+    range_array_appender(begin_t begin, end_t end) : _begin(begin), _end(end) {
+    }
+
+    void operator()(bsoncxx::builder::stream::array_context<> ac) const {
+        for (auto iter = _begin; iter != _end; ++iter) {
+            ac << *iter;
+        }
+    }
+
+   private:
+    begin_t _begin;
+    end_t _end;
+};
+
+template <typename begin_t, typename end_t>
+range_array_appender<begin_t, end_t> make_range_array_appender(begin_t&& begin, end_t&& end) {
+    return range_array_appender<begin_t, end_t>(std::forward<begin_t>(begin),
+                                                std::forward<end_t>(end));
+}
+
+// concatenates arbitrary ranges into an key context
+template <typename begin_t, typename end_t>
+class range_kvp_appender {
+   public:
+    range_kvp_appender(begin_t begin, end_t end) : _begin(begin), _end(end) {
+    }
+
+    void operator()(bsoncxx::builder::stream::key_context<> ac) const {
+        for (auto iter = _begin; iter != _end; ++iter) {
+            ac << iter->first << iter->second;
```

Comment:
So, the requirements on *begin_t and _end_t is that they offer members `first` and `second`, so this would only work with things that dereferce to `std::pair`. I guess that is fine. It wouldn't work with a `vector<bsoncxx::builder::basic::kvp>` or `vector<std::tuple<std::string, int>>`, but you could write a different callable for that, right?